### ISSUE TITLE
`accumulate_grad_batches` is silently ignored when epoch length < `accumulate_grad_batches`, because Lightning always steps on the last batch

### ISSUE THREAD

**Issue by giacomo-ciro** [Lightning-AI/pytorch-lightning#21808]
### Bug description

When a training epoch contains fewer batches than `accumulate_grad_batches`, gradient
accumulation is silently disabled: the optimizer steps on **every** last batch instead
of every `accumulate_grad_batches`-th one. `trainer.global_step` then advances faster.

---

This comes from `TrainingEpochLoop._should_accumulate`:

```python
def _accumulated_batches_reached(self) -> bool:
    """Determine if accumulation will be finished by the end of the current batch."""
    return self.batch_progress.current.ready % self.trainer.accumulate_grad_batches == 0

def _num_ready_batches_reached(self) -> bool:
    """Checks if we are in the last batch or if there are more batches to follow."""
    epoch_finished_on_ready = self.batch_progress.current.ready == self.trainer.num_training_batches
    return epoch_finished_on_ready or self.batch_progress.is_last_batch

def _should_accumulate(self) -> bool:
    """Checks if the optimizer step should be performed or gradients should be accumulated for the current step."""
    accumulation_done = self._accumulated_batches_reached()
    # Lightning steps on the final batch
    is_final_batch = self._num_ready_batches_reached()
    # but the strategy might not
    strategy_accumulates_on_final_batch = self.trainer.strategy.handles_gradient_accumulation or not is_final_batch
    return not accumulation_done and strategy_accumulates_on_final_batch
```

`# Lightning steps on the final batch` is a deliberate design choice (from #2853) meant
to stop accumulation from stalling when an epoch ends mid-cycle (e.g. `accumulate=4` but
only 3 batches left). The unintended consequence: if the epoch is shorter than
`accumulate_grad_batches` in total, *every* batch is the "final batch," so this rule
fires on every step and accumulation never actually happens for the whole run, not
just at the boundary.

This edge case doesn't appear to be documented anywhere (checked the `Trainer`
docstring for `accumulate_grad_batches` and `val_check_interval`, and the gradient
accumulation guide).

### What version are you seeing the problem on?

v2.6

### How to reproduce the bug

```python
``python
import torch
import pytorch_lightning as pl
from torch.utils.data import DataLoader, TensorDataset

class Model(pl.LightningModule):
    def __init__(self):
        super().__init__()
        self.layer = torch.nn.Linear(4, 1)
        self.val_calls = []

    def training_step(self, batch, batch_idx):
        x, y = batch
        return ((self.layer(x) - y) ** 2).mean()

    def validation_step(self, batch, batch_idx):
        self.val_calls.append(self.global_step)
        return torch.tensor(0.0)

    def configure_optimizers(self):
        return torch.optim.SGD(self.parameters(), lr=0.01)

def run(n_train_samples, batch_size, accumulate_grad_batches, val_every_n_steps, max_steps):
    torch.manual_seed(0)
    train_ds = TensorDataset(torch.randn(n_train_samples, 4), torch.randn(n_train_samples, 1))
    val_ds = TensorDataset(torch.randn(20, 4), torch.randn(20, 1))
    train_dl = DataLoader(train_ds, batch_size=batch_size, shuffle=True)
    val_dl = DataLoader(val_ds, batch_size=batch_size)

    model = Model()
    trainer = pl.Trainer(
        max_steps=max_steps,
        accumulate_grad_batches=accumulate_grad_batches,
        val_check_interval=val_every_n_steps * accumulate_grad_batches,  # docs: int = training batches
        check_val_every_n_epoch=None,
        num_sanity_val_steps=0,
        enable_progress_bar=False,
        enable_checkpointing=False,
        enable_model_summary=False,
        logger=False,
    )
    trainer.fit(model, train_dl, val_dl)
    print(f"n_train_samples={n_train_samples:<3} batches/epoch={len(train_dl):<3} "
          f"-> global_step={model.global_step:<4} val fired at steps={sorted(set(model.val_calls))}")

print("Expected in both cases: global_step reaches 40, validation fires at steps [5,10,...,40]\n")

# Epoch (4 batches) is >= accumulate_grad_batches (4): works as documented.
run(n_train_samples=16, batch_size=4, accumulate_grad_batches=4, val_every_n_steps=5, max_steps=40)

# Epoch (1 batch) is < accumulate_grad_batches (4): accumulation is silently
# defeated because every batch is "the last batch of the epoch".
run(n_train_samples=1, batch_size=4, accumulate_grad_batches=4, val_every_n_steps=5, max_steps=40)


Expected in both cases: global_step reaches 40, validation fires at steps [5,10,...,40]

n_train_samples=16  batches/epoch=4   -> global_step=40   val fired at steps=[5, 10, 15, 20, 25, 30, 35, 40]
n_train_samples=1   batches/epoch=1   -> global_step=40   val fired at steps=[20, 40]


With a 16-sample dataset (4 batches/epoch, `accumulate_grad_batches=4`), one epoch =
one accumulation cycle, and everything lines up as documented.

With a 1-sample dataset (1 batch/epoch, same `accumulate_grad_batches=4`),
`global_step` still reaches 40 in the same number of raw batches, but only because
every batch triggered its own optimizer step — i.e. `accumulate_grad_batches=4` had
zero effect. Validation, which was configured to fire every 5 *optimizer* steps, only
fired at step 20 and 40 instead of 5, 10, 15, 20, ....
```

### Error messages and logs

No error is raised or warned. The run completes silently with the wrong effective
batch size and step cadence, which is what makes this easy to miss.

### Environment

<details>
  <summary>Current environment</summary>

```
#- PyTorch Lightning Version (e.g., 2.6.0):  2.6.1
#- PyTorch Version (e.g., 2.5): 2.12.0+cu130
#- Python version (e.g., 3.12): 3.10.9
#- OS (e.g., Linux): Linux-6.8.0-124-generic-x86_64-with-glibc2.39
#- CUDA/cuDNN version: 13.2
#- GPU models and configuration: nVidia RTX 5090
#- How you installed Lightning(`conda`, `pip`, source): uv pip install
```

</details>

### More info

We hit this in practice training on a deliberately tiny (single-object, duplicated to
just 1 sample) dataset with `accumulate_grad_batches=4`: `val_check_interval` computed
as `val_every_n_steps * accumulate_grad_batches` only fired 4x less often than
configured, and it took a while to trace back to this "steps on the final batch" rule
in `_should_accumulate` rather than a bug in our own step-to-batch conversion.

Would it be reasonable to either:
1. document this interaction on `accumulate_grad_batches` / `val_check_interval`, or
2. emit a `PossibleUserWarning` when `num_training_batches < accumulate_grad_batches`,
   since accumulation is silently a no-op in that case?

Happy to send a PR for either if there's agreement on the approach.

cc @ethanwharris

**Comment by deependujha** [comment]
hey @giacomo-ciro, happy to take a PR.

**Comment by giacomo-ciro** [comment]
@deependujha great, and thanks for the quick reply!

Which approach do you prefer, the warning or the docs update?

**Comment by deependujha** [comment]
I'd prefer a warning :) But upto you to decide what you'd preferred.

**Comment by giacomo-ciro** [comment]
Hi @deependujha , the PR #21811 is ready for review :)

**Comment by giacomo-ciro** [comment]
Hi @deependujha , I opened the PR #21811 and I just went with a warning as you suggested. Feel free to review and merge it when you have time. Can't wait to contribute again to this amazing project!

**Comment by deependujha** [comment]
Hi @giacomo-ciro, sorry for the delay, looking into it.

**Comment by giacomo-ciro** [comment]
> Hi [@giacomo-ciro](https://github.com/giacomo-ciro), sorry for the delay, looking into it.

Great, let me know if there is something else I can do.
